### PR TITLE
Mapped stuff payments

### PR DESCRIPTION
**Pull request by AusteZ**
Orders changes so payment status is enum
Payments

### CODE REVIEW

File: Controllers/PaymentsController.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        }
+        [HttpPost("card")]
+        public ActionResult PayWithCard(int orderId, [FromBody] CardPayment card)
+        {
+            return Ok(_service.PayWithCard(_mapper.Map<OrderOutput>(_orderService.Get(orderId)), card));
+        }
+        [HttpPost("cash")]
+        public ActionResult PayWithCash(int orderId)
+        {
+            return Ok(_service.PayWithCash(_mapper.Map<OrderOutput>(_orderService.Get(orderId))));
```
Comment (Rytis-Sapka) [inline Controllers/PaymentsController.cs:33]: Seems like these are returning Orders, is there a way to get the receipt?

File: Services/PaymentsService.cs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+                PaymentType = type,
+            };
+            _receiptRepository.Add(receipt);
+            var receiptOutput = _mapper.Map<ReceiptOutput>(receipt);
+            receiptOutput.Order = order;
+            receiptOutput.PaymentType = type.ToString();
+
+            return receiptOutput;
+        }
+
```
Comment (Rytis-Sapka) [inline Services/PaymentsService.cs:70]: redundant newline

**Review by Rytis-Sapka — CHANGES_REQUESTED** [review]
Looks good, tik pora questions ir pora formatting stuff

File: Services/PaymentsService.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            //To connect some payment gateway
+        }
+        private ReceiptOutput SaveReceipt(OrderOutput order, PaymentType type)
+        {
+            order.Status = PaymentStatus.paid;
+
+            _orderService.Update(_mapper.Map<OrderCreate>(order), order.Id);
+
+            var receipt = new Receipt() {
```
[the remark this answers is not in the document]
Comment (AusteZ) [reply Services/PaymentsService.cs:57]: Have no idea how, all four values are pulling from different sources or are not named the same

File: Controllers/PaymentsController.cs (hunk 1 above)
Comment (AusteZ) [reply Controllers/PaymentsController.cs:33]: PayWithCash method itself returns a receipt, but I will put in get functions to get them after the payment is completed

**Comment by AusteZ** [comment]
Didn't fix with AutoMapper and did not implement Update(Order o) also

File: Controllers/PaymentsController.cs [hunk 4]
```diff
@@ -16,21 +16,34 @@ public class PaymentsController : Controller
...
         private readonly ICrudEntityService<Order, OrderCreate> _orderService;
         private readonly IMapper _mapper;
+
         public PaymentsController(IPaymentService service, ICrudEntityService<Order, OrderCreate> orderService, IMapper mapper)
         {
             _service = service;
             _orderService = orderService;
             _mapper = mapper;
         }
         [HttpPost("card")]
```
Comment (Rytis-Sapka) [inline Controllers/PaymentsController.cs:26]: Add newLines before methods (whole file)

File: Controllers/PaymentsController.cs [hunk 5]
```diff
@@ -16,21 +16,34 @@ public class PaymentsController : Controller
...
         [HttpPost("cash")]
-        public ActionResult PayWithCash(int orderId)
+        public ActionResult PayWithCash(int orderId, float tip)
+        {
+            return Ok(_service.PayWithCash(_mapper.Map<OrderOutput>(_orderService.Get(orderId)), tip));
+        }
+        [HttpGet("Receipt/{orderId}")]
+        public ActionResult GetOrderReceipt(int orderId)
+        {
+            if (_orderService.Get(orderId).Status == PaymentStatus.completed)
```
Comment (Rytis-Sapka) [inline Controllers/PaymentsController.cs:39]: logic should be in services, not controller

File: Services/PaymentsService.cs [hunk 6]
```diff
@@ -18,46 +16,47 @@ public class PaymentsService : IPaymentService
...
                 throw new UserFriendlyException("The order is already paid for", 400);
 
             ProcessPayment();
 
-            return SaveReceipt(order, PaymentType.card);
-            
+            return SaveReceipt(order, PaymentType.card, tip);
         }
-        public ReceiptOutput PayWithCash(OrderOutput order)
+        public ReceiptOutput PayWithCash(OrderOutput order, float tip)
```
Comment (Rytis-Sapka) [inline Services/PaymentsService.cs:41]: new lines before methods (whole file)

File: Services/PaymentsService.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Rytis-Sapka) [reply Services/PaymentsService.cs:55]: ah okay, my bad, did loot take a thorough look

File: Models/Entities/Receipt.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+﻿namespace PSP.Models.Entities
+{
+    public class Receipt
+    {
+        public int OrderId { get; set; }
+        public DateTime? Date { get; set; }
+        public decimal Total { get; set; }
+        public decimal Tip { get; set; }
+        public PaymentType PaymentType { get; set; }
+
```
Comment (Rytis-Sapka) [inline Models/Entities/Receipt.cs:10]: reduntant empty line.

File: Models/Entities/Receipt.cs [hunk 8]
```diff
@@ -0,0 +1,17 @@
+﻿namespace PSP.Models.Entities
+{
+    public class Receipt
+    {
+        public int OrderId { get; set; }
```
Comment (Rytis-Sapka) [inline Models/Entities/Receipt.cs:5]: Missing relational property publid Order Order { get; set; } Will be hard to make a receipt with related products and services without it.

File: Models/DTOs/Output/ReceiptOutput.cs [hunk 9]
```diff
@@ -0,0 +1,11 @@
+﻿namespace PSP.Models.DTOs.Output
+{
+    public class ReceiptOutput
```
Comment (Rytis-Sapka) [inline Models/DTOs/Output/ReceiptOutput.cs:3]: As we discussed, receipts should have related products and services. (or do you want to add this in another PR?)

File: Services/PaymentsService.cs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            _receiptRepository = repository;
+            _orderService = ordersService;
+            _mapper = mapper;
+        }
+
+        public ReceiptOutput PayWithCard(OrderOutput order, CardPayment card, float tip )
+        {
+            if (!new CreditCardAttribute().IsValid(card.CardNumber))
+                throw new UserFriendlyException("The card number is not valid", 400);
+            if (new Regex(@"^[0-9]{3,4}$").IsMatch(card.CVC))
```
Comment (Rytis-Sapka) [inline Services/PaymentsService.cs:31]: this now throws if its valid

**Review by Rytis-Sapka — APPROVED** [review]

**Pull request merged**